pacoloco.go

```diff
@@ -14,17 +14,42 @@ import (
 	"strings"
 	"sync"
 	"time"
+
+	"gorm.io/gorm"
 )
 
 var configFile = flag.String("config", "/etc/pacoloco.yaml", "Path to config file")
+
 var pathRegex *regexp.Regexp
+var filenameRegex *regexp.Regexp   //to get the details of a package (arch, version etc)
+var filenameDBRegex *regexp.Regexp // to get the filename from the db file
+var urlRegex *regexp.Regexp        // to extract the relevant parts from an url to compose a pacoloco url
+
+var prefetchDB *gorm.DB
+
+//Accepted formats
+var allowedPackagesExtensions []string
 
 func init() {
 	var err error
 	pathRegex, err = regexp.Compile("^/repo/([^/]*)(/.*)?/([^/]*)$")
-	if err != nil {
-		panic(err)
-	}
+	//source: https://archlinux.org/pacman/makepkg.conf.5.html PKGEXT
+	allowedPackagesExtensions = []string{".pkg.tar.gz", ".pkg.tar.bz2", ".pkg.tar.xz", ".pkg.tar.zst", ".pkg.tar.lzo", ".pkg.tar.lrz", ".pkg.tar.lz4", ".pkg.tar.lz", ".pkg.tar.Z", ".pkg.tar"}
+	panicIfError(err)
+	//This is ugly, explanation here https://regex101.com/r/qB0fQ7/32
+	filenameRegex, err = regexp.Compile("^([a-z0-9._+-]+)-([^-]+-[0-9]+)-([^-]+)(([.]pkg[.]tar(([.]gz)|([.]bz2)|([.]xz)|([.]zst)|([.]lzo)|([.]lrz)|([.]lz4)|([.]lz)|([.]Z))?)([.]sig)?)$")
```

Comment:
yeah, these regex are a bit hard to follow. In golang it is sometimes more preferable to do a straight-forward (though sometimes a bit more verbose) solution.

I wonder if it is better just use `strings.HasSuffix()` method to match the extension. Once it is matched - remove it, and split the rest using a regexp or just `strings.Split()`.

Also does this code handle package epochs?

Reply:
I had thought about being quite strict on what to accept as a string to prefetch, to avoid things getting out of hand (e.g. path traversal/ weird stuff to access files outside of their bounds).
Now I have added a more complete explanation about how do they are matched.
HasSuffix is being used when I check if a file is a signature or a db one. If it is not one of those, I'll check it against the regex

Reply:
Right now I do not know which are the full list of the allowed characters for the package name (this one should be complete, tested against all the dbs in 3 repos, but if I do not find a documentation about it, it won't be bulletproof)
and especially the epoch/version parameter.
The epoch and versions does not get parsed, they are simply treated as strings cause so far I haven't seen a standard formatting. 
I assume the version in the upstream mirror is the most recent one (also cause the prefetching mechanism should mirror what is being kept upstream, otherwise it would be in an incoherent/detached state wrt the db files served by upstream mirrors)